tests/providers/test_twindb_aws.py

```diff
@@ -24,3 +27,138 @@ def test_main():
+def test_start_instance(mock_boto3, response, expected_code):
+    mock_ec2 = mock.Mock()
+    mock_boto3.resource.return_value = mock_ec2
+    mock_instance = mock.Mock()
+    mock_ec2.instances.filter.return_value = mock_instance
+    mock_instance.start.return_value = response
+    assert start_instance('foo-bar') == expected_code
+    mock_boto3.resource.assert_called_once_with('ec2')
+    mock_ec2.instances.filter.assert_called_once_with(InstanceIds=['foo-bar'])
+    mock_instance.start.assert_called_once()
+
+
+@pytest.mark.parametrize('response, expected_code', [
+    (
+        [{
+            'ResponseMetadata': {
+                'HTTPStatusCode': 200
+            }
+        }],
+        True
+    ),
+    (
+        [{
+            'ResponseMetadata': {
+                'HTTPStatusCode': 404
+            }
+        }],
+        False
+    )
+])
+@mock.patch('twindb_infrastructure.providers.aws.boto3')
+def test_terminate_instance(mock_boto3, response, expected_code):
+    mock_ec2 = mock.Mock()
+    mock_boto3.resource.return_value = mock_ec2
+    mock_instance = mock.Mock()
+    mock_ec2.instances.filter.return_value = mock_instance
+    mock_instance.terminate.return_value = response
+    assert terminate_instance('foo-bar') == expected_code
+    mock_boto3.resource.assert_called_once_with('ec2')
+    mock_ec2.instances.filter.assert_called_once_with(InstanceIds=['foo-bar'])
+    mock_instance.terminate.assert_called_once()
+
+
+@pytest.mark.parametrize('response, expected_code', [
+    (
+        [{
+            'ResponseMetadata': {
+                'HTTPStatusCode': 200
+            }
+        }],
+        True
+    ),
+    (
+        [{
+            'ResponseMetadata': {
+                'HTTPStatusCode': 404
+            }
+        }],
+        False
+    )
+])
+@mock.patch('twindb_infrastructure.providers.aws.boto3')
+def test_stop_instance(mock_boto3, response, expected_code):
+    mock_ec2 = mock.Mock()
+    mock_boto3.resource.return_value = mock_ec2
+    mock_instance = mock.Mock()
+    mock_ec2.instances.filter.return_value = mock_instance
+    mock_instance.stop.return_value = response
+    assert stop_instance('foo-bar') == expected_code
+    mock_boto3.resource.assert_called_once_with('ec2')
+    mock_ec2.instances.filter.assert_called_once_with(InstanceIds=['foo-bar'])
+    mock_instance.stop.assert_called_once()
+
+
+@mock.patch('twindb_infrastructure.providers.aws.boto3')
+def test_start_instance_exception(mock_boto3):
+    mock_boto3.resource.side_effect = ResourceNotExistsError(mock.Mock(), 'error', '')
+    assert not start_instance('foo-bar')
+    mock_boto3.resource.side_effect = UnknownAPIVersionError(mock.Mock(), 'error', '')
+    assert not start_instance('foo-bar')
+    mock_ec2 = mock.Mock()
+    mock_boto3.resource.return_value = mock_ec2
+    mock_instance = mock.Mock()
+    mock_ec2.instances.filter.return_value = mock_instance
+    mock_ec2.instances.side_effect = ClientError
```

Comment:
if you look at coverage report you'll see lines 313-314 are not covered.
https://codecov.io/gh/twindb/twindb_infrastructure/pull/8/src/twindb_infrastructure/providers/aws.py

I would suggest to split this test into four and test each case separately:

1. `resource()` raises `ResourceNotExistsError`
2. `resource()` raises `UnknownAPIVersionError` 
3.  `ec2.instances.filter()` raises `ClientError`
4.  `instance.start()` raises `ClientError`